### DISCUSSION TITLE
Stream audio and video in one connection

### DISCUSSION THREAD

**Discussion by Shchusia** [aiortc/aiortc#953] [Q&A]
Greetings.
I encountered a problem when trying to transfer video and audio from an IP camera. When debugging, I discovered that frames(video+audio) are sent from the client but the server does not receive them. When I try to stream only audio or video, everything works correctly.
On the server. I initialize the connection like this:
```python
peer_connection = RTCPeerConnection(
        configuration=RTCConfiguration(
            iceServers=[
                RTCIceServer(**url_data.dict()) for url_data in CONFIG.sturn_turn_config
            ]
        )
    )
recorder = MediaRecorder(str(path/file_name))
@peer_connection.on("track")
async def on_track(track):
        peer_connection.addTrack(track)
        recorder.addTrack(track)

       @track.on("ended")
       async def on_ended():
            print("Track ended", track.kind)
            await recorder.stop()

peer_connection.addTransceiver("video")
peer_connection.addTransceiver('audio')
offer = await peer_connection.createOffer()
await peer_connection.setLocalDescription(offer)
```

publisher initialization looks like this
```python
offer = RTCSessionDescription(sdp=sdp_server["sdp"], type=sdp_server["type"])
    peer_connection = RTCPeerConnection(
        configuration=RTCConfiguration(
            iceServers=[
                RTCIceServer(**url_data.model_dump()) for url_data in sturn_turn_config
            ]
        )
    )
await peer_connection.setRemoteDescription(offer)
mp = MediaPlayer(**params)
# I tried without MediaRelay
relay = MediaRelay()
stream_v = relay.subscribe(media_player.video)
stream_a = relay.subscribe(media_player.audio)
peer_connection.addTrack(stream_v)
peer_connection.addTrack(stream_a)
answer = await peer_connection.createAnswer()
await peer_connection.setRemoteDescription(offer)
await peer_connection.setLocalDescription(answer)
descr = peer_connection.localDescription
```
 
In this configuration, if you use only one type of transmission or video or audio, everything works, but not all together.
Tell me what could be the problem.

**Comment by qnlbnsl** [comment]
Hey, I was trying to figure out how to stream audio only. Do you have the code that worked?

**Reply by vinzbarbuto** [reply]
Hello! To exclusively stream audio, begin by acquiring the audio through the following steps:

```python
player = MediaPlayer('default:default', format='avfoundation', options=options)
audio_stream = player.audio
``` 

The format parameter, in this case 'avfoundation', may vary depending on the device you are extracting audio from. I recommend consulting the [FFmpeg Devices List](https://ffmpeg.org/ffmpeg-devices.html) to identify the appropriate format and explore available options.

After obtaining the audio, seamlessly integrate it into the peer connection's track with the following code:
```python
pc.addTrack(audio_stream)
``` 
This ensures that the audio stream is properly added to the peer connection for streamlined audio-only streaming.

**Comment by vinzbarbuto** [comment]
Hi, I have the same problem. Did you find a solution?

**Reply by IsmaelDeveloper** [reply]
same here. did you find a solution ?

**Reply by vinzbarbuto** [reply]
Unfortunately, I haven't found a definitive solution yet. As a temporary workaround, I'm currently employing two separate peer connections—one for video and another for audio. While this isn't an ideal or elegant fix, it does manage to get the job done for now.

**Reply by IsmaelDeveloper** [reply]
I see thank you ! :(

**Reply by Shchusia** [reply]
I abandoned the self-written stream server and started using Janus. The problem was on the server side with aiortc, it doesn’t want to work, but transferring data to Janus in one connection(audio/video) works. also works on the front side with one connection.

I did the same with 2 connections. but another problem was identified. MediaRecorder disrupts the stream, which causes delays + when trying to make a scaled system, the quality also drops.
